Source/Urho3D/IO/ChunkStreamDeserializer.cpp

```diff
@@ -0,0 +1,261 @@
+    auto* chunk = FindChunk(position);
+    if (chunk)
+    {
+        if (deserializer_.Seek(chunk->position_) != chunk->position_)
+            return -1;
+        if (!ReadBlock(deserializer_, chunk->decodedSize_, chunk->size_))
+            return -1;
+        readBufferSize_ = chunk->decodedSize_;
+        readBufferOffset_ = position - chunk->decodedPosition_;
+        return position;
+    }
+
+    // Seek beyond last known chunk.
+    chunk = GetLastChunk();
+    if (!chunk)
+    {
+        position_ = 0;
+        readBufferOffset_ = 0;
+        readBufferSize_ = 0;
+        if (deserializer_.Seek(offset_) != offset_)
+        {
+            return -1;
+        }
+    }
+    else
+    {
+        position_ = chunk->decodedPosition_ + chunk->decodedSize_;
+        if (deserializer_.Seek(chunk->position_ + chunk->size_) != chunk->position_ + chunk->size_)
+        {
+            return -1;
+        }
+    }
+
+    // Read chunks until we reach end of file or find a chunk where data belongs to.
+    for (;;)
+    {
+        if (deserializer_.IsEof())
+        {
+            if (position_ == position)
+                return position;
+            return -1;
+        }
+        unsigned short unpackedSize = deserializer_.ReadUShort();
+        unsigned short packedSize = deserializer_.ReadUShort();
+        const auto chunkStartPos = deserializer_.GetPosition();
+        Chunk newChunk{ chunkStartPos, packedSize, position_, unpackedSize };
+        Add(newChunk);
+
+        // Check if we reached destination.
+        if (position < position_ + unpackedSize)
+        {
+            if (!ReadBlock(deserializer_, unpackedSize, packedSize))
+                return -1;
+            readBufferSize_ = unpackedSize;
+            readBufferOffset_ = position - newChunk.decodedPosition_;
+            return position;
+        }
+
+        position_ += unpackedSize;
+        if (deserializer_.Seek(newChunk.position_ + newChunk.size_) != newChunk.position_ + newChunk.size_)
+        {
+            return -1;
+        }
+
+    }
+
+    return position_;
+}
+
+/// Register new chunk. All chunk should be sequential in position.
+bool ChunkStreamDeserializer::Add(const Chunk& chunk)
+{
+    int index = chunks_.size();
+    if (index > 0)
+    {
+        if (chunks_[index - 1].position_ >= chunk.position_ ||
+            chunks_[index - 1].decodedPosition_ >= chunk.decodedPosition_)
+        {
+            return false;
+        }
+    }
+    chunks_.push_back(chunk);
+    return true;
+}
+
+/// Find chunk by stream position.
+/// Returns chunk index if found or -1 otherwise.
+const ChunkStreamDeserializer::Chunk* ChunkStreamDeserializer::FindChunk(unsigned decodedPosition) const
+{
+    // Early out for an empty collection.
+    if (chunks_.empty())
+        return nullptr;
+
+    // Check if position we looking for is beyond last chunk start.
+    const Chunk& lastChunk = chunks_.back();
+    if (decodedPosition >= lastChunk.decodedPosition_)
```

Comment:
Is there a point in this code block? Doesnt same thing after binary search take care of this case?

Reply:
Actually it revealed a bug I had :)

Reply:
It would be good to add a comment about this bug. It clearly is not obvious and next time someone looks at this code they may think of cleaning it up (just like i thought of it) and break it without being aware of some cornercase